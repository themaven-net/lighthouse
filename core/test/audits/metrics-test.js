/**
 * @license Copyright 2018 The Lighthouse Authors. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
 */

import jestMock from 'jest-mock';

import MetricsAudit from '../../audits/metrics.js';
import {Interactive} from '../../computed/metrics/interactive.js';
import {getURLArtifactFromDevtoolsLog, readJson} from '../test-utils.js';
import {defaultSettings} from '../../config/constants.js';

const pwaTrace = readJson('../fixtures/traces/progressive-app-m60.json', import.meta);
const pwaDevtoolsLog = readJson('../fixtures/traces/progressive-app-m60.devtools.log.json', import.meta);
const lcpTrace = readJson('../fixtures/traces/lcp-m78.json', import.meta);
const lcpDevtoolsLog = readJson('../fixtures/traces/lcp-m78.devtools.log.json', import.meta);
const lcpImageTrace = readJson('../fixtures/traces/amp-m86.trace.json', import.meta);
const lcpImageDevtoolsLog = readJson('../fixtures/traces/amp-m86.devtoolslog.json', import.meta);
const lcpAllFramesTrace = readJson('../fixtures/traces/frame-metrics-m89.json', import.meta);
const lcpAllFramesDevtoolsLog = readJson('../fixtures/traces/frame-metrics-m89.devtools.log.json', import.meta);
const clsAllFramesTrace = readJson('../fixtures/traces/frame-metrics-m90.json', import.meta);
const clsAllFramesDevtoolsLog = readJson('../fixtures/traces/frame-metrics-m90.devtools.log.json', import.meta);
const jumpyClsTrace = readJson('../fixtures/traces/jumpy-cls-m90.json', import.meta);
const jumpyClsDevtoolsLog = readJson('../fixtures/traces/jumpy-cls-m90.devtoolslog.json', import.meta);

const settings = JSON.parse(JSON.stringify(defaultSettings));

describe('Performance: metrics', () => {
  it('evaluates valid input correctly', async () => {
    const URL = getURLArtifactFromDevtoolsLog(pwaDevtoolsLog);
    const artifacts = {
      URL,
      GatherContext: {gatherMode: 'navigation'},
      traces: {
        [MetricsAudit.DEFAULT_PASS]: pwaTrace,
      },
      devtoolsLogs: {
        [MetricsAudit.DEFAULT_PASS]: pwaDevtoolsLog,
      },
    };

    const context = {
      settings: {...settings, throttlingMethod: 'simulate'},
      computedCache: new Map(),
    };
    const result = await MetricsAudit.audit(artifacts, context);
    expect(result.details.items[0]).toMatchSnapshot();
  });

  it('evaluates valid input correctly (throttlingMethod=provided)', async () => {
    const URL = getURLArtifactFromDevtoolsLog(pwaDevtoolsLog);
    const artifacts = {
      URL,
      GatherContext: {gatherMode: 'navigation'},
      traces: {
        [MetricsAudit.DEFAULT_PASS]: pwaTrace,
      },
      devtoolsLogs: {
        [MetricsAudit.DEFAULT_PASS]: pwaDevtoolsLog,
      },
    };

    const context = {
      settings: {...settings, throttlingMethod: 'provided'},
      computedCache: new Map(),
    };
    const result = await MetricsAudit.audit(artifacts, context);
    expect(result.details.items[0]).toMatchSnapshot();
  });

  it('evaluates valid input (with lcp) correctly', async () => {
    const URL = getURLArtifactFromDevtoolsLog(lcpDevtoolsLog);
    const artifacts = {
      URL,
      GatherContext: {gatherMode: 'navigation'},
      traces: {
        [MetricsAudit.DEFAULT_PASS]: lcpTrace,
      },
      devtoolsLogs: {
        [MetricsAudit.DEFAULT_PASS]: lcpDevtoolsLog,
      },
    };

    const context = {
      settings: {...settings, throttlingMethod: 'simulate'},
      computedCache: new Map(),
    };
    const result = await MetricsAudit.audit(artifacts, context);
    expect(result.details.items[0]).toMatchSnapshot();
  });

  it('evaluates valid input (with lcp from all frames) correctly', async () => {
    const URL = getURLArtifactFromDevtoolsLog(lcpAllFramesDevtoolsLog);
    const artifacts = {
      URL,
      GatherContext: {gatherMode: 'navigation'},
      traces: {
        [MetricsAudit.DEFAULT_PASS]: lcpAllFramesTrace,
      },
      devtoolsLogs: {
        [MetricsAudit.DEFAULT_PASS]: lcpAllFramesDevtoolsLog,
      },
    };

    const context = {
      settings: {...settings, throttlingMethod: 'provided'},
      computedCache: new Map(),
    };
    const result = await MetricsAudit.audit(artifacts, context);
    expect(result.details.items[0]).toMatchSnapshot();
  });

  it('evaluates valid input (with image lcp) correctly', async () => {
    const URL = getURLArtifactFromDevtoolsLog(lcpImageDevtoolsLog);
    const artifacts = {
      URL,
      GatherContext: {gatherMode: 'navigation'},
      traces: {
        [MetricsAudit.DEFAULT_PASS]: lcpImageTrace,
      },
      devtoolsLogs: {
        [MetricsAudit.DEFAULT_PASS]: lcpImageDevtoolsLog,
      },
    };

    const context = {
      settings: {...settings, throttlingMethod: 'simulate'},
      computedCache: new Map(),
    };
    const result = await MetricsAudit.audit(artifacts, context);
    expect(result.details.items[0]).toMatchSnapshot();
  });

  it('leaves CLS undefined in an old trace without weighted scores', async () => {
    const artifacts = {
      GatherContext: {gatherMode: 'navigation'},
      traces: {
        [MetricsAudit.DEFAULT_PASS]: lcpAllFramesTrace,
      },
      devtoolsLogs: {
        [MetricsAudit.DEFAULT_PASS]: lcpAllFramesDevtoolsLog,
      },
    };

    const context = {
      settings: {...settings, throttlingMethod: 'simulate'},
      computedCache: new Map(),
    };
    const {details} = await MetricsAudit.audit(artifacts, context);
    expect(details.items[0]).toMatchObject({
      cumulativeLayoutShift: undefined,
      cumulativeLayoutShiftMainFrame: undefined,
      totalCumulativeLayoutShift: undefined,
      observedCumulativeLayoutShift: undefined,
      observedCumulativeLayoutShiftMainFrame: undefined,
      observedTotalCumulativeLayoutShift: undefined,
    });
  });

  it('evaluates new CLS correctly across all frames', async () => {
    const URL = getURLArtifactFromDevtoolsLog(clsAllFramesDevtoolsLog);
    const artifacts = {
      URL,
      GatherContext: {gatherMode: 'navigation'},
      traces: {
        [MetricsAudit.DEFAULT_PASS]: clsAllFramesTrace,
      },
      devtoolsLogs: {
        [MetricsAudit.DEFAULT_PASS]: clsAllFramesDevtoolsLog,
      },
    };

    const context = {
      settings: {...settings, throttlingMethod: 'provided'},
      computedCache: new Map(),
    };
    const {details} = await MetricsAudit.audit(artifacts, context);

    // Only a single main-frame shift event, so mfCls and oldCls are equal.
    expect(details.items[0]).toMatchObject({
      cumulativeLayoutShift: expect.toBeApproximately(0.026463, 6),
      cumulativeLayoutShiftMainFrame: expect.toBeApproximately(0.001166, 6),
      totalCumulativeLayoutShift: expect.toBeApproximately(0.001166, 6),

      observedCumulativeLayoutShift: expect.toBeApproximately(0.026463, 6),
      observedCumulativeLayoutShiftMainFrame: expect.toBeApproximately(0.001166, 6),
      observedTotalCumulativeLayoutShift: expect.toBeApproximately(0.001166, 6),
    });
  });

  it('does not fail the entire audit when TTI errors', async () => {
    const URL = getURLArtifactFromDevtoolsLog(pwaDevtoolsLog);
    const artifacts = {
      URL,
      GatherContext: {gatherMode: 'navigation'},
      traces: {
        [MetricsAudit.DEFAULT_PASS]: pwaTrace,
      },
      devtoolsLogs: {
        [MetricsAudit.DEFAULT_PASS]: pwaDevtoolsLog,
      },
    };

    const mockTTIFn = jestMock.spyOn(Interactive, 'request');
    mockTTIFn.mockRejectedValueOnce(new Error('TTI failed'));
    const context = {
      settings: {...settings, throttlingMethod: 'simulate'},
      computedCache: new Map(),
    };
    const result = await MetricsAudit.audit(artifacts, context);
    expect(result.details.items[0].interactive).toEqual(undefined);
  });

  it('evaluates CLS correctly', async () => {
    const URL = getURLArtifactFromDevtoolsLog(jumpyClsDevtoolsLog);
    const artifacts = {
      URL,
      GatherContext: {gatherMode: 'navigation'},
      traces: {
        [MetricsAudit.DEFAULT_PASS]: jumpyClsTrace,
      },
      devtoolsLogs: {
        [MetricsAudit.DEFAULT_PASS]: jumpyClsDevtoolsLog,
      },
    };

    const context = {
      settings: {...settings, throttlingMethod: 'simulate'},
      computedCache: new Map(),
    };
    const {details} = await MetricsAudit.audit(artifacts, context);
    expect(details.items[0]).toMatchObject({
      cumulativeLayoutShift: expect.toBeApproximately(2.268816, 6),
      cumulativeLayoutShiftMainFrame: expect.toBeApproximately(2.268816, 6),
      totalCumulativeLayoutShift: expect.toBeApproximately(4.809794, 6),

      observedCumulativeLayoutShift: expect.toBeApproximately(2.268816, 6),
      observedCumulativeLayoutShiftMainFrame: expect.toBeApproximately(2.268816, 6),
      observedTotalCumulativeLayoutShift: expect.toBeApproximately(4.809794, 6),
    });
  });
});
