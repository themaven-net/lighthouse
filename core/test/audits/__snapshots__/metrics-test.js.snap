// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Performance: metrics evaluates valid input (with image lcp) correctly 1`] = `
Object {
  "cumulativeLayoutShift": 0,
  "cumulativeLayoutShiftMainFrame": 0,
  "firstContentfulPaint": 3364,
  "firstContentfulPaintAllFrames": undefined,
  "firstContentfulPaintAllFramesTs": undefined,
  "firstContentfulPaintTs": undefined,
  "firstMeaningfulPaint": 3364,
  "firstMeaningfulPaintTs": undefined,
  "interactive": 6412,
  "interactiveTs": undefined,
  "largestContentfulPaint": 5578,
  "largestContentfulPaintAllFrames": undefined,
  "largestContentfulPaintAllFramesTs": undefined,
  "largestContentfulPaintTs": undefined,
  "lcpLoadEnd": 5529,
  "lcpLoadStart": 5407,
  "maxPotentialFID": 160,
  "observedCumulativeLayoutShift": 0,
  "observedCumulativeLayoutShiftMainFrame": 0,
  "observedDomContentLoaded": 2474,
  "observedDomContentLoadedTs": 760623117637,
  "observedFirstContentfulPaint": 2685,
  "observedFirstContentfulPaintAllFrames": 2685,
  "observedFirstContentfulPaintAllFramesTs": 760623328111,
  "observedFirstContentfulPaintTs": 760623328111,
  "observedFirstMeaningfulPaint": 2685,
  "observedFirstMeaningfulPaintTs": 760623328111,
  "observedFirstPaint": 2685,
  "observedFirstPaintTs": 760623328111,
  "observedFirstVisualChange": 2652,
  "observedFirstVisualChangeTs": 760623295599,
  "observedLargestContentfulPaint": 2882,
  "observedLargestContentfulPaintAllFrames": 2882,
  "observedLargestContentfulPaintAllFramesTs": 760623525330,
  "observedLargestContentfulPaintTs": 760623525330,
  "observedLastVisualChange": 3269,
  "observedLastVisualChangeTs": 760623912599,
  "observedLoad": 2697,
  "observedLoadTs": 760623340349,
  "observedNavigationStart": 0,
  "observedNavigationStartTs": 760620643599,
  "observedSpeedIndex": 2795,
  "observedSpeedIndexTs": 760623438456,
  "observedTimeOrigin": 0,
  "observedTimeOriginTs": 760620643599,
  "observedTotalCumulativeLayoutShift": 0,
  "observedTraceEnd": 4778,
  "observedTraceEndTs": 760625421283,
  "speedIndex": 6339,
  "speedIndexTs": undefined,
  "timeToFirstByte": 2394,
  "timeToFirstByteTs": undefined,
  "totalBlockingTime": 162,
  "totalCumulativeLayoutShift": 0,
}
`;

exports[`Performance: metrics evaluates valid input (with lcp from all frames) correctly 1`] = `
Object {
  "cumulativeLayoutShift": undefined,
  "cumulativeLayoutShiftMainFrame": undefined,
  "firstContentfulPaint": 863,
  "firstContentfulPaintAllFrames": 683,
  "firstContentfulPaintAllFramesTs": 23466705983,
  "firstContentfulPaintTs": 23466886143,
  "firstMeaningfulPaint": 863,
  "firstMeaningfulPaintTs": 23466886143,
  "interactive": 863,
  "interactiveTs": 23466886143,
  "largestContentfulPaint": 863,
  "largestContentfulPaintAllFrames": 683,
  "largestContentfulPaintAllFramesTs": 23466705983,
  "largestContentfulPaintTs": 23466886143,
  "lcpLoadEnd": undefined,
  "lcpLoadStart": undefined,
  "maxPotentialFID": 16,
  "observedCumulativeLayoutShift": undefined,
  "observedCumulativeLayoutShiftMainFrame": undefined,
  "observedDomContentLoaded": 596,
  "observedDomContentLoadedTs": 23466619325,
  "observedFirstContentfulPaint": 863,
  "observedFirstContentfulPaintAllFrames": 683,
  "observedFirstContentfulPaintAllFramesTs": 23466705983,
  "observedFirstContentfulPaintTs": 23466886143,
  "observedFirstMeaningfulPaint": 863,
  "observedFirstMeaningfulPaintTs": 23466886143,
  "observedFirstPaint": 616,
  "observedFirstPaintTs": 23466639588,
  "observedFirstVisualChange": 609,
  "observedFirstVisualChangeTs": 23466632130,
  "observedLargestContentfulPaint": 863,
  "observedLargestContentfulPaintAllFrames": 683,
  "observedLargestContentfulPaintAllFramesTs": 23466705983,
  "observedLargestContentfulPaintTs": 23466886143,
  "observedLastVisualChange": 5881,
  "observedLastVisualChangeTs": 23471904130,
  "observedLoad": 673,
  "observedLoadTs": 23466696096,
  "observedNavigationStart": 0,
  "observedNavigationStartTs": 23466023130,
  "observedSpeedIndex": 1583,
  "observedSpeedIndexTs": 23467605703,
  "observedTimeOrigin": 0,
  "observedTimeOriginTs": 23466023130,
  "observedTotalCumulativeLayoutShift": undefined,
  "observedTraceEnd": 6006,
  "observedTraceEndTs": 23472029453,
  "speedIndex": 1583,
  "speedIndexTs": 23467606130,
  "timeToFirstByte": 565,
  "timeToFirstByteTs": 23466588051,
  "totalBlockingTime": 0,
  "totalCumulativeLayoutShift": undefined,
}
`;

exports[`Performance: metrics evaluates valid input (with lcp) correctly 1`] = `
Object {
  "cumulativeLayoutShift": 0,
  "cumulativeLayoutShiftMainFrame": 0,
  "firstContentfulPaint": 2289,
  "firstContentfulPaintAllFrames": undefined,
  "firstContentfulPaintAllFramesTs": undefined,
  "firstContentfulPaintTs": undefined,
  "firstMeaningfulPaint": 2758,
  "firstMeaningfulPaintTs": undefined,
  "interactive": 4367,
  "interactiveTs": undefined,
  "largestContentfulPaint": 2758,
  "largestContentfulPaintAllFrames": undefined,
  "largestContentfulPaintAllFramesTs": undefined,
  "largestContentfulPaintTs": undefined,
  "lcpLoadEnd": undefined,
  "lcpLoadStart": undefined,
  "maxPotentialFID": 1336,
  "observedCumulativeLayoutShift": 0,
  "observedCumulativeLayoutShiftMainFrame": 0,
  "observedDomContentLoaded": 1513,
  "observedDomContentLoadedTs": 713038536140,
  "observedFirstContentfulPaint": 1122,
  "observedFirstContentfulPaintAllFrames": 1122,
  "observedFirstContentfulPaintAllFramesTs": 713038144775,
  "observedFirstContentfulPaintTs": 713038144775,
  "observedFirstMeaningfulPaint": 1122,
  "observedFirstMeaningfulPaintTs": 713038144775,
  "observedFirstPaint": 1122,
  "observedFirstPaintTs": 713038144775,
  "observedFirstVisualChange": 1105,
  "observedFirstVisualChangeTs": 713038128064,
  "observedLargestContentfulPaint": 1122,
  "observedLargestContentfulPaintAllFrames": 1122,
  "observedLargestContentfulPaintAllFramesTs": 713038144775,
  "observedLargestContentfulPaintTs": 713038144775,
  "observedLastVisualChange": 1722,
  "observedLastVisualChangeTs": 713038745064,
  "observedLoad": 2159,
  "observedLoadTs": 713039182071,
  "observedNavigationStart": 0,
  "observedNavigationStartTs": 713037023064,
  "observedSpeedIndex": 1393,
  "observedSpeedIndexTs": 713038416494,
  "observedTimeOrigin": 0,
  "observedTimeOriginTs": 713037023064,
  "observedTotalCumulativeLayoutShift": 0,
  "observedTraceEnd": 7416,
  "observedTraceEndTs": 713044439102,
  "speedIndex": 3681,
  "speedIndexTs": undefined,
  "timeToFirstByte": 609,
  "timeToFirstByteTs": undefined,
  "totalBlockingTime": 1205,
  "totalCumulativeLayoutShift": 0,
}
`;

exports[`Performance: metrics evaluates valid input correctly (throttlingMethod=provided) 1`] = `
Object {
  "cumulativeLayoutShift": 0,
  "cumulativeLayoutShiftMainFrame": 0,
  "firstContentfulPaint": 499,
  "firstContentfulPaintAllFrames": 499,
  "firstContentfulPaintAllFramesTs": 225414670885,
  "firstContentfulPaintTs": 225414670885,
  "firstMeaningfulPaint": 783,
  "firstMeaningfulPaintTs": 225414955343,
  "interactive": 1582,
  "interactiveTs": 225415754204,
  "largestContentfulPaint": undefined,
  "largestContentfulPaintAllFrames": undefined,
  "largestContentfulPaintAllFramesTs": undefined,
  "largestContentfulPaintTs": undefined,
  "lcpLoadEnd": undefined,
  "lcpLoadStart": undefined,
  "maxPotentialFID": 198,
  "observedCumulativeLayoutShift": 0,
  "observedCumulativeLayoutShiftMainFrame": 0,
  "observedDomContentLoaded": 560,
  "observedDomContentLoadedTs": 225414732309,
  "observedFirstContentfulPaint": 499,
  "observedFirstContentfulPaintAllFrames": 499,
  "observedFirstContentfulPaintAllFramesTs": 225414670885,
  "observedFirstContentfulPaintTs": 225414670885,
  "observedFirstMeaningfulPaint": 783,
  "observedFirstMeaningfulPaintTs": 225414955343,
  "observedFirstPaint": 499,
  "observedFirstPaintTs": 225414670868,
  "observedFirstVisualChange": 520,
  "observedFirstVisualChangeTs": 225414692015,
  "observedLargestContentfulPaint": undefined,
  "observedLargestContentfulPaintAllFrames": undefined,
  "observedLargestContentfulPaintAllFramesTs": undefined,
  "observedLargestContentfulPaintTs": undefined,
  "observedLastVisualChange": 818,
  "observedLastVisualChangeTs": 225414990015,
  "observedLoad": 2199,
  "observedLoadTs": 225416370913,
  "observedNavigationStart": 0,
  "observedNavigationStartTs": 225414172015,
  "observedSpeedIndex": 605,
  "observedSpeedIndexTs": 225414776724,
  "observedTimeOrigin": 0,
  "observedTimeOriginTs": 225414172015,
  "observedTotalCumulativeLayoutShift": 0,
  "observedTraceEnd": 12540,
  "observedTraceEndTs": 225426711887,
  "speedIndex": 605,
  "speedIndexTs": 225414777015,
  "timeToFirstByte": 261,
  "timeToFirstByteTs": 225414432704,
  "totalBlockingTime": 48,
  "totalCumulativeLayoutShift": 0,
}
`;

exports[`Performance: metrics evaluates valid input correctly 1`] = `
Object {
  "cumulativeLayoutShift": 0,
  "cumulativeLayoutShiftMainFrame": 0,
  "firstContentfulPaint": 1337,
  "firstContentfulPaintAllFrames": undefined,
  "firstContentfulPaintAllFramesTs": undefined,
  "firstContentfulPaintTs": undefined,
  "firstMeaningfulPaint": 1541,
  "firstMeaningfulPaintTs": undefined,
  "interactive": 4205,
  "interactiveTs": undefined,
  "largestContentfulPaint": undefined,
  "largestContentfulPaintAllFrames": undefined,
  "largestContentfulPaintAllFramesTs": undefined,
  "largestContentfulPaintTs": undefined,
  "lcpLoadEnd": undefined,
  "lcpLoadStart": undefined,
  "maxPotentialFID": 396,
  "observedCumulativeLayoutShift": 0,
  "observedCumulativeLayoutShiftMainFrame": 0,
  "observedDomContentLoaded": 560,
  "observedDomContentLoadedTs": 225414732309,
  "observedFirstContentfulPaint": 499,
  "observedFirstContentfulPaintAllFrames": 499,
  "observedFirstContentfulPaintAllFramesTs": 225414670885,
  "observedFirstContentfulPaintTs": 225414670885,
  "observedFirstMeaningfulPaint": 783,
  "observedFirstMeaningfulPaintTs": 225414955343,
  "observedFirstPaint": 499,
  "observedFirstPaintTs": 225414670868,
  "observedFirstVisualChange": 520,
  "observedFirstVisualChangeTs": 225414692015,
  "observedLargestContentfulPaint": undefined,
  "observedLargestContentfulPaintAllFrames": undefined,
  "observedLargestContentfulPaintAllFramesTs": undefined,
  "observedLargestContentfulPaintTs": undefined,
  "observedLastVisualChange": 818,
  "observedLastVisualChangeTs": 225414990015,
  "observedLoad": 2199,
  "observedLoadTs": 225416370913,
  "observedNavigationStart": 0,
  "observedNavigationStartTs": 225414172015,
  "observedSpeedIndex": 605,
  "observedSpeedIndexTs": 225414776724,
  "observedTimeOrigin": 0,
  "observedTimeOriginTs": 225414172015,
  "observedTotalCumulativeLayoutShift": 0,
  "observedTraceEnd": 12540,
  "observedTraceEndTs": 225426711887,
  "speedIndex": 1676,
  "speedIndexTs": undefined,
  "timeToFirstByte": 759,
  "timeToFirstByteTs": undefined,
  "totalBlockingTime": 777,
  "totalCumulativeLayoutShift": 0,
}
`;
